Create 2023-11-28-server-recon-explained.md (#145)
swimos/swimos-docs: src/_posts/2023-11-28-server-recon-explained.md

# Server Recon Explained

> `server-recon` is a configuration file format used by the SwimOS platform.
> As the name suggests, it uses <a href="http://docs.swim.ai/js/latest/modules/_swim_recon.html" target="_target">Recon notation</a>, which looks likes an extension of JSON.

Reviewer: ```suggestion
<a href="http://docs.swim.ai/js/latest/modules/_swim_recon.html" target="_target">Recon</a> is object notation used SwimOS platform for configuration files and for communication between Web Agents.
```

The are three primary parts of SwimOS server configration:

> - web attributes
> - fabric attributes
> - kernel directives

Reviewer: ```suggestion
- Web attributes
- Fabric attributes
- Kernel directives
```

## Web Attributes

The `@web` annotation is used to specify Web Attributes. 
There is a top-level field called `port` that specific the TCP/IP port being used.
There is a field called `space` the corresponds to an application and maps to `swim.api.space.Space`.
The value of this field will be the same name used when defining the fabric attribute.

> There is a field called `documentRoot` that optionally specifies the top-level UI directory when applicable.

Reviewer: ```suggestion
There is a field called `documentRoot` that optionally specifies the top-level UI directory for bundled UIs, when applicable.
```

There is an attribute called `@websocket` that allows `serverCompressionLevel` and `clientCompressionLevel` to be specfied.

> Server compression corresponds to the Web Agent side, while client corresponds to the use of client libraries, such as through ` swim.client.ClientRuntime`.

Reviewer: ```suggestion

```

```
  @node {
    pattern: "/city/:id"
    @agent(class: "swim.yourspace.agent.CityAgent")
  }
```

Here, `id` is a route variable used to uniquely identify a given `CityAgent`.

The second approach consists of specify the underlying web address for a specific agent. When using this, it is customary to specify a `seed` field for the corresponding `@agent` attribute that populates a `seed` field with configuration info, such as the name of a file. `swim.api.agent.AbstractAgent.getProp("seed")` can then be invoked on the Web Agent to retrieve the configuration information.

In same cases, it may be desirable to specify multiple `@agent` attributes for the same configuration. This allows a component approach that separates concerns, such as having both a core agent and a logging agent. What happens is that each Web Agent implementation is matched against `@SwimLane` defined on each Web Agent, and each corresponding lane match is invoked. As a result, multiple Web Agents may run concurrently as a single Web Agent entity.

## Kernel Directives

> 

Author: @SirCipher fixed: https://github.com/swimos/swimos-docs/pull/145/commits/3c803c77dd44d512650d31d11583c64bc68afc8e
